File: ios/HybridNitroGoogleSso.swift
import Foundation
import NitroModules
import GoogleSignIn
import UIKit

class HybridNitroGoogleSso: HybridNitroGoogleSsoSpec {
    let domain: String = "com.nitro.google.sso"
    let nitroGoogleSSOImpl = NitroGoogleSSOImpl()
    
    func configure(config: NitroGoogleSSOConfig) throws {
        GIDSignIn.sharedInstance.configuration = nitroGoogleSSOImpl.getConfig(config)
    }
    
    func getCurrentUser() throws -> Promise<NitroGoogleUserInfo?> {
        return .async { [weak self] in
            guard let self else {
                throw NitroGoogleSSOError.selfNoAvailable
            }
            do {
                return try await self.nitroGoogleSSOImpl.getCurrentUser()
            } catch {
                throw error
            }
        }
    }
    
    func signIn() throws -> Promise<NitroGoogleUserInfo?> {
        return .async { [weak self] in
            guard let self else {
                throw NitroGoogleSSOError.selfNoAvailable
            }
            do {
                return try await nitroGoogleSSOImpl.signIn()
            } catch {
                throw error
            }
        }
    }
    
    func signOut() throws -> Promise<Void> {
        return .async { [weak self] in
            guard let self else {
                throw NitroGoogleSSOError.selfNoAvailable
            }
            do {
                try await self.nitroGoogleSSOImpl.signOut()
            } catch {
                throw error
            }
        }
    }
}

File: android/src/main/java/com/nitrogooglesso/HybridNitroGoogleSso.kt
package com.nitrogooglesso

import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitrogooglesso.HybridNitroGoogleSsoSpec
import com.margelo.nitro.nitrogooglesso.NitroGoogleSSOConfig
import com.margelo.nitro.nitrogooglesso.NitroGoogleUserInfo
import com.margelo.nitro.nitrogooglesso.Variant_NullType_NitroGoogleUserInfo
import kotlinx.coroutines.MainScope

class HybridNitroGoogleSso: HybridNitroGoogleSsoSpec() {
    val context = NitroModules.applicationContext ?: throw Error("HybridNitroGoogleSSO context not found")
    val mainScope = MainScope()
    lateinit var nitroGoogleSSOImpl: NitroGoogleSSOImpl

    override fun configure(config: NitroGoogleSSOConfig) {
        nitroGoogleSSOImpl = NitroGoogleSSOImpl(context, config)
    }

    override fun signIn(): Promise<Variant_NullType_NitroGoogleUserInfo> {
        return Promise.async(mainScope) {
            nitroGoogleSSOImpl.signIn()
        }
    }

    override fun oneTapSignIn(): Promise<Variant_NullType_NitroGoogleUserInfo> {
        return Promise.async(mainScope) {
            nitroGoogleSSOImpl.oneTapSignIn()
        }
    }

    override fun signOut(): Promise<Unit> {
        return Promise.async(mainScope) {
            nitroGoogleSSOImpl.signOut()
        }
    }

    override fun getCurrentUser(): Promise<Variant_NullType_NitroGoogleUserInfo> {
        return Promise.async(mainScope) {
            nitroGoogleSSOImpl.getCurrentUser()
        }
    }
}
